### PR TITLE
Add support for AES  encryption & decryption.

### PR DESCRIPTION
**Pull request by xqdoo00o**
It's just basic code, but surely work for zip aes decryption & encryption. I'm not sure the code flow is enough good, and the 3rd library([sjcl](https://github.com/bitwiseshiftleft/sjcl)) may need async require(or not, min.js only 20kb bigger).
But if the file size large than [DEFAULT_BLOCK_SIZE](https://github.com/Stuk/jszip/blob/master/lib/stream/DataWorker.js#L8), I think there is a bug , because the data was split.

### CODE REVIEW

**Comment by xqdoo00o** [comment]
Hi @Stuk, could you please review this?

File: lib/zipEntry.js [hunk 1]
```diff
@@ -118,9 +130,9 @@ ZipEntry.prototype = {
         this.externalFileAttributes = reader.readInt(4);
         this.localHeaderOffset = reader.readInt(4);
 
-        if (this.isEncrypted()) {
-            throw new Error("Encrypted zip are not supported");
-        }
+        // if (this.isEncrypted()) {
+        //     throw new Error("Encrypted zip are not supported");
+        // }
```
Comment (Stuk) [inline lib/zipEntry.js:135]: Please remove this commented code.

File: lib/zipEntry.js [hunk 2]
```diff
@@ -203,13 +215,21 @@ ZipEntry.prototype = {
...
             extraFieldLength = reader.readInt(2);
-            extraFieldValue = reader.readData(extraFieldLength);
 
-            this.extraFields[extraFieldId] = {
-                id: extraFieldId,
-                length: extraFieldLength,
-                value: extraFieldValue
-            };
+            // AES encrypt extra field header id
+            if(extraFieldId === 0x9901){
```
Comment (Stuk) [inline lib/zipEntry.js:220]:
```suggestion
            if (extraFieldId === 0x9901){
```

File: lib/zipEntry.js [hunk 3]
```diff
@@ -203,13 +215,21 @@ ZipEntry.prototype = {
...
+                this.options.aes.version = reader.readInt(2);
+                reader.skip(2);
+                this.options.aes.strength = reader.readInt(1);
+                this.compressionMethod = this.options.aes.compressionMethod = reader.readString(2);
+            }else{
```
Comment (Stuk) [inline lib/zipEntry.js:225]:
```suggestion
            } else {
```

Please also change the other places where this does not follow the existing convention.

**Comment by xqdoo00o** [comment]
OK, I tested.
origin min.js 94kb -- 27kb gziped
withCrypto min.js 117KB -- 34kb gziped
New commit include test & format fix has been pushed.

File: lib/stream/AesWorker.js [hunk 4]
```diff
@@ -60,56 +70,47 @@ AesWorker.prototype.processChunk = function (chunk) {
...
-        iterations: 1000
-    }).toString();
-    var aesKey = this._crypto.enc.Hex.parse(derivedKey.slice(0, this._keyLen * 2));
-    var macKey = this._crypto.enc.Hex.parse(derivedKey.slice(this._keyLen * 2, (this._keyLen + this._macLen) * 2));
-    var derivedPassVerifier = derivedKey.slice(-this._passVerifyLen * 2);
-    if (this._aesAction === "Encrypt") {
-        derivedPassVerifier = this._crypto.enc.Hex.parse(derivedPassVerifier);
+    var hmac;
+    var iv = [0, 0, 0, 0];
+    var salt = this._aesAction === "Encrypt" ? this._crypto.random.randomWords(this._saltLen / 4) : this._crypto.codec.bytes.toBits(data.slice(0, this._saltLen));
```
Comment (deliangyang) [inline lib/stream/AesWorker.js:75]: The length of this line is too long
```suggestion
var salt = this._aesAction === "Encrypt" 
	? this._crypto.random.randomWords(this._saltLen / 4) 
	: this._crypto.codec.bytes.toBits(data.slice(0, this._saltLen));
```

File: lib/ctrGladman.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+      word += b3;	  
+		} else {
+		  word += (0x01 << 24);
+		}
+		return word;
+	},
+
+	incCounter: function(counter) {
+		if ((counter[0] = this.incWord(counter[0])) === 0)
+		{
```
Comment (deliangyang) [inline lib/ctrGladman.js:80]:
```suggestion
		if ((counter[0] = this.incWord(counter[0])) === 0) {
```

File: lib/ctrGladman.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+   * @throws {sjcl.exception.invalid} if the IV isn't exactly 128 bits or if any adata is specified.
+   * @throws {sjcl.exception.corrupt} if if the message is corrupt.
+   */
+  decrypt: function(prf, ciphertext, iv, adata) {
+    return sjcl.mode.ctrGladman._calculate(prf, ciphertext, iv, adata);
+  },
+
+  incWord: function(word) {	
+		if (((word >> 24) & 0xff) === 0xff) { //overflow
+      var b1 = (word >> 16) & 0xff;
```
Comment (deliangyang) [inline lib/ctrGladman.js:48]: indentation needing

**Comment by xqdoo00o** [comment]
I pull the new efficient AES encrypt/decrypt code in https://github.com/Stuk/jszip/pull/696

**Pull request closed without merging**